analyzers/src/SonarAnalyzer.CSharp/Rules/LoopsAndLinq.cs

```diff
@@ -43,15 +48,96 @@ public sealed class LoopsAndLinq : SonarDiagnosticAnalyzer
         protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
                 {
-                    var node = (ForEachStatementSyntax)c.Node;
-
-                    if (node.Statement is BlockSyntax blockSyntax
-                        && blockSyntax.ChildNodes().Count() == 1
-                        && blockSyntax.ChildNodes().Single() is IfStatementSyntax)
+                    var forEachStatementSyntax = (ForEachStatementSyntax)c.Node;
+                    if (CanBeSimplifiedUsingWhere(forEachStatementSyntax.Statement))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, forEachStatementSyntax.Expression.GetLocation(), WhereMessageFormat));
+                    }
+                    else
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, node.GetLocation()));
+                        CheckIfCanBeSimplifiedUsingSelect(forEachStatementSyntax, c);
                     }
                 },
                 SyntaxKind.ForEachStatement);
+
+        private static bool CanBeSimplifiedUsingWhere(SyntaxNode statement) =>
+            GetIfStatement(statement) is { } ifStatementSyntax
+            && CanIfStatementBeMoved(ifStatementSyntax);
+
+        private static IfStatementSyntax GetIfStatement(SyntaxNode node) =>
+            node switch
+            {
+                IfStatementSyntax ifStatementSyntax => ifStatementSyntax,
+                BlockSyntax blockSyntax when blockSyntax.ChildNodes().Count() == 1 => GetIfStatement(blockSyntax.ChildNodes().Single()),
+                _ => null
+            };
+
+        private static bool CanIfStatementBeMoved(IfStatementSyntax ifStatementSyntax) =>
+            ifStatementSyntax.Else == null
+            && !ContainsExitPoints(ifStatementSyntax)
+            && ifStatementSyntax.Condition is InvocationExpressionSyntax invocationExpressionSyntax
+            && !invocationExpressionSyntax.DescendantNodes().OfType<ArgumentSyntax>().Any(argument => argument.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword));
+
+        private static bool ContainsExitPoints(SyntaxNode syntaxNode) =>
+            syntaxNode.DescendantNodes()
+                      .Any(node => node.IsAnyKind(SyntaxKind.ReturnStatement,
+                                                  SyntaxKind.YieldReturnStatement,
```

Comment:
ThrowStatementSyntax, ContinueStatementSyntax

what about containing an `await` inside the `if` block?

Reply:
I've added `ThrowStatementSyntax` and `ContinueStatementSyntax`. I think `await` is not really changing the flow and we can skip it for now.

Reply:
And, in the end, I've decided to skip this early exit point check. It didn't bring any value.